karapace/kafka_rest_apis/__init__.py

```diff
@@ -523,16 +524,23 @@ async def cluster_metadata(self, topics: Optional[List[str]] = None) -> dict:
             if self._metadata_birth is None or time.monotonic() - self._metadata_birth > self.metadata_max_age:
                 self._cluster_metadata = None
 
-            if self._cluster_metadata and topics is None:
-                return self._cluster_metadata
+            if self._cluster_metadata:
+                # Return from metadata only if all queried topics have cached metadata
+                if topics is None:
+                    if self._cluster_metadata_complete:
+                        return self._cluster_metadata
+                elif all(topic in self._cluster_metadata["topics"] for topic in topics):
+                    return {
+                        **self._cluster_metadata,
+                        "topics": {topic: self._cluster_metadata["topics"][topic] for topic in topics},
+                    }
 
             try:
                 metadata_birth = time.monotonic()
                 metadata = self.admin_client.cluster_metadata(topics)
-                # Only store cache if fetched full metadata
-                if topics is None:
-                    self._metadata_birth = metadata_birth
-                    self._cluster_metadata = metadata
+                self._metadata_birth = metadata_birth
+                self._cluster_metadata = metadata
+                self._cluster_metadata_complete = topics is None
             except NodeNotReadyError:
                 log.exception("Could not refresh cluster metadata")
                 KafkaRest.r(
```

Comment:
Note, gateway error would be better here than internal server error.

Reply:
Maybe address this separately, and see if there is anything similar in CSR from compatibility point of view?

Reply:
Sure, just notices this when reviewing and added as note.